src/utilities/codemirror/custom-folder.ts

```diff
@@ -129,3 +138,234 @@ function foldMetadataOrModel(
     to: newEnd,
   };
 }
+
+type BlockStackNode = Readonly<{
+  node: SyntaxNode;
+  stem: string;
+}>;
+
+type BlockStack = BlockStackNode[];
+
+export type PairedCommands = {
+  end: SyntaxNode;
+  endPos: number;
+  start: SyntaxNode;
+  startPos: number;
+};
+
+type PartialPairedCommands = Partial<PairedCommands>;
+
+type TreeState = {
+  [startPos: number]: PartialPairedCommands;
+};
+
+export function isPairedCommands(pair: unknown): pair is PairedCommands {
+  const pc = pair as PairedCommands;
+  return !!pc?.start && !!pc?.end;
+}
+
+type BlockType = Readonly<{
+  close: string;
+  open: string[];
+  partition?: string;
+}>;
+
+const SEQ_DIR_END_IF = 'SEQ_DIR_END_IF';
```

Comment:
Do you think these constants can be potentially reused outside of this file? If so, do you mind moving them to a constants file under `/constants`?

Reply:
I agree, but may we defer that part?
I want to minimize merge collisions over the next week as we pull in the other demo items

Reply:
No problem! I created a ticket to track it https://github.com/NASA-AMMOS/aerie-ui/issues/1470